### ISSUE TITLE
Error becomes large for long sequence in multiHeadCrossAttentionPlugin

### ISSUE THREAD

**Issue by 1049451037** [NVIDIA/TensorRT#2674]
## Description

I'm trying to use [multiHeadCrossAttentionPlugin](https://github.com/NVIDIA/TensorRT/tree/release/8.5/plugin/multiHeadCrossAttentionPlugin) in my engine. However, I find the deviation (drift) of its output to standard attention becomes larger as the sequence length becoming longer.

Concretely, the max absolute error for the output of fMHCA and standard attention (written by pytorch) is as follows:

![var](https://user-images.githubusercontent.com/15194939/218364023-c75279a5-2b6f-4f9d-ac32-2992e52850b3.png)

x axis is the sequence length, y axis is max abs error for two outputs. Batch size is 2, number of head is 8, head size is 40. Inputs are generated with standard normal distribution. It seems that only for 64 and 128 seq_len, the error is tolerable.

## Environment

nvidia docker container 22.12

**Comment by zerollzeng** [comment]
Can you provide a reproduce for this issue? Thanks!
cc @ttyio

**Comment by 1049451037** [comment]
Here is the python code to reproduce:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.onnx import OperatorExportTypes
import os
import tensorrt as trt
from torch.testing._internal.common_utils import numpy_to_torch_dtype_dict

class fMHCA(torch.autograd.Function):
    @staticmethod
    def forward(ctx, q, kv):
        """
        q: (batch_size, seq_len, num_head, size_per_head)
        kv: (batch_size, seq_len, num_head, 2, size_per_head)
        output: like q
        """
        size_per_head = q.size(3)
        batch_size = q.size(0)
        seq_len = q.size(1)
        num_head = q.size(2)
        q = q.transpose(1, 2).contiguous()
        kv = kv.transpose(1, 2).contiguous()
        k = kv.select(-2, 0)
        v = kv.select(-2, 1)
        scores = torch.matmul(q, k.transpose(-2, -1)) * (size_per_head**-0.5)
        scores = F.softmax(scores, -1)
        result = torch.matmul(scores, v).transpose(1, 2).contiguous().view(batch_size, seq_len, num_head, size_per_head)
        return result
    @staticmethod
    def symbolic(g, q, kv):
        return g.op("fMHCA", q, kv, plugin_version_s='1')

class CrossAttn(nn.Module):
    def forward(self, q, kv):
        return fMHCA.apply(q, kv)

TRT_LOGGER = trt.Logger()
trt.init_libnvinfer_plugins(TRT_LOGGER, '')

def load_engine(engine_file_path):
    assert os.path.exists(engine_file_path)
    print("Reading engine from file {}".format(engine_file_path))
    with open(engine_file_path, "rb") as f, trt.Runtime(TRT_LOGGER) as runtime:
        return runtime.deserialize_cuda_engine(f.read())

def generate_trt(model, inputs, kw_args={}, experiment_name=''):
    os.makedirs('./onnx/', exist_ok=True)
    os.makedirs('./trt/', exist_ok=True)
    name = type(model).__name__ + '_' + experiment_name
    model.eval()
    with torch.no_grad():
        outputs = model(*inputs, **kw_args)
        num_output = 1 if type(outputs) is not list and type(outputs) is not tuple else len(outputs)
        os.system("rm onnx/{}.onnx".format(name))
        torch.onnx.export(model, tuple(inputs)+(kw_args,), 'onnx/{}.onnx'.format(name), operator_export_type=OperatorExportTypes.ONNX_FALLTHROUGH, input_names=['input_{}'.format(i) for i in range(len(inputs)+len(kw_args))], output_names=['output_{}'.format(i) for i in range(num_output)])
    os.system("rm trt/{}.trt".format(name))
    os.system("trtexec --onnx=onnx/{}.onnx --saveEngine=trt/{}.trt --fp16 --buildOnly".format(name, name))
    return 'trt/'+name+'.trt'

def inference_trt(engine_path, inputs_dict):
    engine = load_engine(engine_path)
    context = engine.create_execution_context()
    outputs_dict = {}
    bindings = []
    for binding in engine:
        binding_idx = engine.get_binding_index(binding)
        # size = trt.volume(context.get_binding_shape(binding_idx))
        dtype = trt.nptype(engine.get_binding_dtype(binding))
        if engine.binding_is_input(binding):
            bindings.append(int(inputs_dict[binding].data_ptr()))
        else:
            shape = tuple(context.get_binding_shape(binding_idx))
            outputs_dict[binding] = torch.empty(*shape, dtype=numpy_to_torch_dtype_dict[dtype], device='cuda')
            bindings.append(int(outputs_dict[binding].data_ptr()))
    stream = torch.cuda.default_stream()
    context.execute_async_v2(bindings=bindings, stream_handle=stream.cuda_stream)
    stream.synchronize()
    return outputs_dict['output_0']

def experiment(seq_len):
    q = torch.randn(2, seq_len, 8, 40).half().cuda()
    kv = torch.randn(2, seq_len, 8, 2, 40).half().cuda()
    model = CrossAttn().cuda()
    trt_name = generate_trt(model, (q, kv))
    o1 = model(q, kv)
    o2 = inference_trt(trt_name, {'input_0': q, 'input_1': kv})
    return (o1-o2).abs().max().item()

def main():
    x = []
    y = []
    for l in range(64, 1024+1, 64):
        x.append(l)
        v = experiment(l)
        y.append(v)
    import matplotlib.pyplot as plt
    plt.scatter(x, y)
    plt.savefig('var.png')

if __name__ == '__main__':
    main()
```

**Comment by 1049451037** [comment]
By the way, it will be better if attention mask is available for cross-attention...

**Comment by zerollzeng** [comment]
@ttyio @samurdhikaru Do you think it's a known issue or can we make some improvement on this? Thanks!

**Comment by ttyio** [comment]
Hi @1049451037 , for the mask enhancement, there is no plan so far since this is opensource plugin, feel free to modify the code `cuSeqLens` in your project https://github.com/NVIDIA/TensorRT/blob/main/plugin/multiHeadCrossAttentionPlugin/fmhcaPlugin.cpp#L99

For the accuracy issue, will look into this, thanks!

**Comment by 1049451037** [comment]
I think mask enhancement is necessary. When key and value are text information (which often are), we can only pass batch size as 1 if no mask support.

**Comment by ttyio** [comment]
@1049451037 for the accuracy issue, pytorch use fp32 acc for softmax (`softmax_warp_forward<c10::Half, c10::Half, float, ...`), and the trt fp16 plugin use fp16 acc. That explains why accuracy drops with larger sequence. 
Thanks!

**Comment by ttyio** [comment]
@1049451037 , the accuracy issue is fixed in 8.6 OOTB solution (without using the plugin), could you take a try? thank you!

**Comment by 1049451037** [comment]
Thank you for your work. However, I'm using nvidia docker container to run TensorRT. And I notice that there is no available 8.6 docker. (Moreover, I don't have access of cuda 12.0 machine to run the latest docker...) Sorry...

**Comment by ttyio** [comment]
We have official 8.6 docker, I will close this, pls reopen if you still have question, thanks!
